### PR TITLE
Add delegate registration class - Closes #884

### LINKED ISSUE #884: Add DelegateTransaction class
### Description

Extend `BaseTransaction` implement all the specific features of register delegate transactions.

### PR DESCRIPTION
**Pull request by mitsuaki-u**
### Description

Added Delegate Registration class.

* The PR resolves #884 
* All new code is covered with unit tests
* All new code was formatted with Prettier
* Linting passes
* Tests pass
* Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
* Documentation has been added/updated

### CODE REVIEW

File: packages/lisk-transactions/src/transactions/2_register_delegate.ts [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+
+		return transaction;
+	}
+
+	protected getAssetBytes(): Buffer {
+		const {
+			delegate: { username },
+		} = this.asset;
+
+		return username && typeof username === 'string'
```
Comment (shuse2) [inline packages/lisk-transactions/src/transactions/2_register_delegate.ts]: username is required field, so no need to check here

File: packages/lisk-transactions/src/transactions/2_register_delegate.ts [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+			errors.push(
+				new TransactionError(
+					`Username is not unique.`,
+					this.id,
+					'.asset.delegate.username',
+				),
+			);
+		}
+
+		if ((sender as Account).isDelegate || sender.username) {
```
Comment (shuse2) [inline packages/lisk-transactions/src/transactions/2_register_delegate.ts]: i think u don't need type conversion here

File: packages/lisk-transactions/src/transactions/2_register_delegate.ts [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+			errors.push(
+				new TransactionError(
+					'Account is already a delegate',
+					this.id,
+					'.username',
```
Comment (shuse2) [inline packages/lisk-transactions/src/transactions/2_register_delegate.ts]: transaction wise, datapath is `.asset.delegate.username`?

File: packages/lisk-transactions/src/transactions/2_register_delegate.ts [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+		if (!usernameUnique) {
+			errors.push(
+				new TransactionError(
+					`Username is not unique.`,
+					this.id,
+					'.asset.delegate.username',
+				),
+			);
+		}
+		if ((sender as Account).isDelegate || sender.username) {
```
Comment (shuse2) [inline packages/lisk-transactions/src/transactions/2_register_delegate.ts]: same as above

File: packages/lisk-transactions/src/transactions/2_register_delegate.ts [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+			errors.push(
+				new TransactionError(
+					'Account is already a delegate',
+					this.id,
+					'.username',
+				),
+			);
+		}
+		// Ignore state from the base transaction
```
Comment (shuse2) [inline packages/lisk-transactions/src/transactions/2_register_delegate.ts]: i think u don't need to ignore the state, it should be used?

File: packages/lisk-transactions/src/transactions/2_register_delegate.ts [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+		};
+	}
+
+	public undo({ sender }: RequiredDelegateState): TransactionResponse {
+		const { errors: baseErrors, state } = super.undo({ sender });
+		if (!state) {
+			throw new Error('State is required for undoing transaction.');
+		}
+		const errors = [...baseErrors];
+		// Ignore state from the base transaction
```
Comment (shuse2) [inline packages/lisk-transactions/src/transactions/2_register_delegate.ts]: same as above

**Review by shuse2 — DISMISSED** [review]

File: packages/lisk-transactions/src/transactions/2_register_delegate.ts [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+		if (!usernameUnique) {
+			errors.push(
+				new TransactionError(
+					`Username is not unique.`,
+					this.id,
+					'.asset.delegate.username',
+				),
+			);
+		}
+		if (sender.isDelegate || sender.username) {
```
Comment (shuse2) [inline packages/lisk-transactions/src/transactions/2_register_delegate.ts]: No need to consider original sender, i think only considering the updated sender (state.sender) is fine

File: packages/lisk-transactions/src/transactions/2_register_delegate.ts [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+	type: 'object',
+	required: ['delegate'],
+	properties: {
+		delegate: {
+			type: 'object',
+			required: ['username'],
+			properties: {
+				username: {
+					type: 'string',
+					minLength: 1,
```
Comment (shuse2) [inline packages/lisk-transactions/src/transactions/2_register_delegate.ts]: min and max is kind of format, so putting only in format should be fine?

File: packages/lisk-transactions/src/transactions/2_register_delegate.ts [hunk 9]
```diff
@@ -0,0 +1,405 @@
...
+					maxLength: 20,
+					format: 'username',
+				},
+			},
+		},
+	},
+};
+
+export interface CreateDelegateRegistrationInput {
```
Comment (shuse2) [inline packages/lisk-transactions/src/transactions/2_register_delegate.ts:86]: probably better to extend CreateBaseTransactionInput like
`export type CastVoteInput = CreateBaseTransactionInput & CreateVoteAssetInput;`

**Review by shuse2 — DISMISSED** [review]

File: packages/lisk-transactions/src/utils/validation/validation.ts [hunk 10]
```diff
@@ -36,6 +36,24 @@ export const validatePublicKey = (publicKey: string) => {
 	return true;
 };
 
+export const validateUsername = (username: string) => {
+	if (
+		username !== username.toLowerCase()
```
Comment (shuse2) [inline packages/lisk-transactions/src/utils/validation/validation.ts]: here it should be trimmed too

**Review by shuse2 — DISMISSED** [review]

**Review by shuse2 — APPROVED** [review]

File: packages/lisk-transactions/src/transactions/2_register_delegate.ts [hunk 11]
```diff
@@ -0,0 +1,393 @@
...
+			type: 2,
+			fee: DELEGATE_FEE.toString(),
+			asset: { delegate: { username } },
+		};
+
+		if (!passphrase) {
+			return transaction;
+		}
+
+		const transferTransaction = new DelegateTransaction(
```
Comment (SargeKhan) [inline packages/lisk-transactions/src/transactions/2_register_delegate.ts]: It's a delegateTransaction. :)

**Review by SargeKhan — DISMISSED** [review]

**Pull request merged**